lib/hanami/common_logger.rb

```diff
@@ -73,11 +82,17 @@ def log(env, status, header, began_at)
         logger.info(msg)
       end
     end
+    # rubocop:enable Metrics/MethodLength
+    # rubocop:enable Metrics/AbcSize
 
     # @since 1.1.0
     # @api private
     def extract_params(env)
-      (env[QUERY_HASH] || {}).merge(env[FORM_HASH] || {})
+      result = env.fetch(QUERY_HASH, {})
+      result.merge!(env.fetch(FORM_HASH, {}))
+      result.merge!(Utils::Hash.deep_stringify(env.fetch(ROUTER_PARAMS, {})))
+      result.delete(BODY_PARSER_FALLBACK_KEY)
```

Comment:
Just for better understanding: could you explain why we need to drop it? and I understand right that we will drop all `'_'` keys?

Reply:
@davydovanton You made me spot a problem with this. Here's why I Initially dropped `_`. 

---

Rack/Hanami params are a key/value bucket. It may happen that a JSON payload is an array (e.g. `'["a", "b", "c"]'`), so we don't know which key to assign to the array. For this reason, we assign this fallback key for cases like this. (eg. `{ "_" => ["a", "b", "c"]`)

See https://github.com/hanami/router/issues/85 and https://github.com/hanami/router/pull/86

---

When a request has a `Content-Type` equal to `x-www-form-urlencoded` (a web form submission), the JSON body parser doesn't recognize the payload, and it adds it to the fallback key. See https://github.com/hanami/router/issues/179

---

My clumsy attempt here was to avoid the automatic logger to print a messed up log.

```log
 {"book"=>{"title"=>"GO"}, "_"=>"book%5Btitle%5D=GO"}
```

---

But I forgot two things:

  1. There is a PR by @timriley to fix the body parsers in the first place. See https://github.com/hanami/router/pull/180
  1. If we remove `_` from here, it won't log the JSON arrays

---

I'm gonna change this. Thanks. 👍

Reply:
@davydovanton Thanks to https://github.com/hanami/router/pull/181, we don't need to remove the fallback key anymore. See https://github.com/hanami/hanami/pull/958/commits/10b3e455ef097a211cecfb249617d1ccdb4efdc7